core/pta/attestation.c

```diff
@@ -0,0 +1,600 @@
+	size_t e_size = 0;
+
+	res = pub_asn1_decode(&n_ptr, &e_ptr, &n_size, &e_size);
+	if (res)
+		return res;
+
+	if (n_ptr == NULL || e_ptr == NULL)
+		return TEE_ERROR_GENERIC;
+
+	// encrypt the random number with the public key from the att data
+	res = crypto_acipher_alloc_rsa_public_key(&key, 0);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(e_ptr, e_size, key.e);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(n_ptr, n_size, key.n);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	size_t enc_rn_len = n_size;
+	uint8_t *enc_rn = calloc(enc_rn_len, sizeof(uint8_t));
+
+	res = crypto_acipher_rsaes_encrypt(TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1,
+					&key,
+					NULL, 0,
+					rn, rn_len,
+					enc_rn, &enc_rn_len);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	crypto_acipher_free_rsa_public_key(&key);
+
+	// use the private key to decrypt the random number
+	struct rsa_keypair key_pair;
+	uint8_t *d_ptr = NULL;
+	size_t d_size = 0;
+	size_t dec_rn_len = n_size+1;
+	uint8_t *dec_rn = calloc(dec_rn_len, sizeof(uint8_t));
+
+	res = crypto_acipher_alloc_rsa_keypair(&key_pair, 0);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = priv_asn1_decode(plain_priv, &d_ptr, &d_size);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(d_ptr, d_size, key_pair.d);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(e_ptr, e_size, key_pair.e);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_bignum_bin2bn(n_ptr, n_size, key_pair.n);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	res = crypto_acipher_rsaes_decrypt(TEE_ALG_RSAES_PKCS1_OAEP_MGF1_SHA1,
+					&key_pair,
+					NULL, 0,
+					enc_rn, enc_rn_len,
+					dec_rn, &dec_rn_len);
+	if (res)
+		return TEE_ERROR_SECURITY;
+
+	// validate the decrypted random number vs the original one
+	for (size_t i = 0; i < dec_rn_len; i++) {
+		if (rn[i] != dec_rn[i]) {
+			DMSG("%s failed!", __func__);
+			return TEE_ERROR_SECURITY;
+		}
+	}
+
+	free(plain_priv);
+	free(enc_rn);
+	free(dec_rn);
+	crypto_bignum_free(key_pair.e);
+	crypto_bignum_free(key_pair.d);
+	crypto_bignum_free(key_pair.n);
+	return res;
+}
+
+/*
+ * The set data function gets the attestation data from the normal world.
+ * This data is verified and saved. The data is device specific and used to
+ * forge the attestation certificate.
+ */
+static TEE_Result set_data(uint32_t type, TEE_Param p[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_INPUT,
+					  TEE_PARAM_TYPE_MEMREF_INPUT);
+	if (exp_pt != type)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	att_data.iv_size = p[0].memref.size;
+	att_data.pub_size = p[1].memref.size;
+	att_data.enc_priv_size = p[2].memref.size;
+	att_data.sig_size = p[3].memref.size;
+
+	// Allocating the memory for the Attestation data
+	att_data.iv = calloc(att_data.iv_size, sizeof(uint8_t));
```

Comment:
Memory leak if `set_data()` is called several times?

Reply:
You are right, I will address that

Reply:
Fix [here](https://github.com/OP-TEE/optee_os/pull/4011/commits/d3129d491a63be354fb9eb9d83b108285b3e384b)